feat: Allow to disable in-cluster server address. Fixes #7571

terrytangyuan opened the pull request:
Fixes https://github.com/argoproj/argo-cd/issues/7571.

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

jessesuen commented on util/settings/settings.go:
```diff
@@ -323,6 +325,8 @@ const (
 	// settingsPasswordPatternKey is the key to configure user password regular expression
 	settingsPasswordPatternKey = "passwordPattern"
+	// inClusterServerAddressAllowedKey is the key to configure whether to allow in-cluster server address
+	inClusterServerAddressAllowedKey = "inClusterServerAddressAllowed"
```
Let's nest this under a `cluster` settings namespace (like we do with repository, application, resource, etc...):

```yaml
data:
  cluster.inClusterEnabled: false
```

jessesuen commented on util/settings/settings.go:
```diff
@@ -1093,6 +1105,7 @@ func updateSettingsFromConfigMap(settings *ArgoCDSettings, argoCDCM *apiv1.Confi
 	if settings.PasswordPattern == "" {
 		settings.PasswordPattern = common.PasswordPatten
 	}
+	settings.InClusterServerAddressAllowed = argoCDCM.Data[inClusterServerAddressAllowedKey] == "true"
```
I think this is a bug. If this setting doesn't exist, it should return true. So this should probably be:

```
settings.InClusterServerAddressAllowed = argoCDCM.Data[clusterInClusterEnabled] != "false"
```

jessesuen commented on util/settings/settings.go:
```diff
@@ -544,6 +548,14 @@ func (mgr *SettingsManager) GetTrackingMethod() (string, error) {
 	return argoCDCM.Data[settingsResourceTrackingMethodKey], nil
 }
 
+func (mgr *SettingsManager) GetInClusterServerAddressAllowed() (bool, error) {
+	argoCDCM, err := mgr.getConfigMap()
+	if err != nil {
+		return true, err
+	}
+	return argoCDCM.Data[inClusterServerAddressAllowedKey] == "true", nil
+}
+
```
For simple settings fields that are just strings and don't need yaml marshalling, we don't need to use accessor methods. So it is okay to delete this method and the code will rely on `settings.clusterInClusterEnabled`.

jessesuen commented on util/db/cluster.go:
```diff
@@ -60,21 +60,31 @@ func (db *db) ListClusters(ctx context.Context) (*appv1.ClusterList, error) {
+			} else {
+				log.Errorf("failed to add cluster %q to cluster list: in-cluster server address is disabled in Argo CD settings", cluster.Name)
+			}
+		} else {
+			if !hasInClusterCredentials {
+				clusterList.Items = append(clusterList.Items, *db.getLocalCluster())
+			}
+			clusterList.Items = append(clusterList.Items, *cluster)
```
I think this is a bug. It seems it keeps adding the cluster over and over again.

terrytangyuan replied:
Updated the logic. Could you take another look when you get a chance?

terrytangyuan commented:
E2E tests failed with the following error:

```
time="2021-11-07T04:13:10Z" level=info msg="expectation succeeded: no error and output contained ''"
time="2021-11-07T04:13:10Z" level=info msg="pending: sync status to be Synced, is "
time="2021-11-07T04:13:13Z" level=info msg="pending: sync status to be Synced, is "
time="2021-11-07T04:13:16Z" level=info msg="pending: sync status to be Synced, is "
time="2021-11-07T04:13:19Z" level=info msg="pending: sync status to be Synced, is "
time="2021-11-07T04:13:22Z" level=info msg="pending: sync status to be Synced, is "
Error:     app_autosync_test.go:28: timeout waiting for: sync status to be Synced, is 
--- FAIL: TestAutoSyncSelfHealDisabled (19.70s)
FAIL
FAIL	github.com/argoproj/argo-cd/v2/test/e2e	29.336s
FAIL
```
Looks like the sync status is empty. Any suggestions on where my changes might have affected that?

jessesuen commented on util/db/cluster.go:
```diff
@@ -60,16 +60,27 @@ func (db *db) ListClusters(ctx context.Context) (*appv1.ClusterList, error) {
+			clusterList.Items = append(clusterList.Items, *cluster)
 		}
 	}
 	if !hasInClusterCredentials {
```
This if check needs to consult if inClusterEnabled is turned off.

terrytangyuan replied:
Good catch! Updated.

jessesuen reviewed (changes requested):
Could we also modify CreateCluster in this file to check if c.Cluster == appv1.KubernetesInternalAPIServerAddr and reject the request?

terrytangyuan commented:
> Could we also modify CreateCluster in this file to check if c.Cluster == appv1.KubernetesInternalAPIServerAddr and reject the request? e.g.:
> 
> ```go
> 	settings, err := db.settingsMgr.GetSettings()
> 	if err != nil {
> 		return nil, err
> 	}
> 	if c.Server == appv1.KubernetesInternalAPIServerAddr && !settings.InClusterEnabled {
> 		return nil, status.Errorf(codes.InvalidArgument, "cannot register cluster: in-cluster has been disabled"))
> 	}
> ```

Updated and added unit tests. Thanks for the feedback!

jessesuen commented:
I tried running the failing test locally and it is passing on my local machine:

```
$ ./hack/test.sh -v -run TestAutoSyncSelfHealDisabled
...
time="2021-11-11T00:44:35-08:00" level=info msg="expectation succeeded: no error and output contained ''"
time="2021-11-11T00:44:35-08:00" level=info msg="pending: sync status to be Synced, is OutOfSync"
time="2021-11-11T00:44:38-08:00" level=info msg="expectation succeeded: sync status to be Synced, is Synced"
time="2021-11-11T00:44:38-08:00" level=info msg="pending: sync status to be OutOfSync, is Synced"
time="2021-11-11T00:44:41-08:00" level=info msg="expectation succeeded: sync status to be OutOfSync, is OutOfSync"
--- PASS: TestAutoSyncSelfHealDisabled (16.45s)
PASS
ok      github.com/argoproj/argo-cd/v2/test/e2e 18.695s
+ report
+ set -eux -o pipefail
+ go-junit-report
```

This is on `v1.21.5+k3s1` k3s

terrytangyuan commented:
This seems to happen randomly in different versions of k3s for e2e tests.

terrytangyuan commented:
Running only the v1.21.2 e2e tests and CI passed https://github.com/argoproj/argo-cd/runs/5110660738?check_suite_focus=true

terrytangyuan commented:
v1.20.2 also failed: https://github.com/argoproj/argo-cd/runs/5111118572?check_suite_focus=true
